053-configurable-event-store/proposal.md

```diff
@@ -0,0 +1,452 @@
+1. I imagine this can be a performance improvement for certain backends. Even
+   with Postgres, every time there's a new build event, we perform a query - I
+   would predict that just reading from the notification bus could be faster
+   (especially considering the current proceduce is: i) receive empty event from
+   notification bus, and ii) perform a query) - this is just a hypothesis, though.
+   That said, other backends (e.g. an S3 backend) could be relatively slow to `Get`.
+
+Note that `EventStore` implementations don't need to think about notifications.
+This all happens at the DB layer:
+
+```go
+func (b *build) SaveEvent(ctx context.Context, event atc.Event) error {
+    // use the `EventStore` to save
+    key, err := b.eventStore.Put(ctx, b, event)
+    if err != nil {
+        return err
+    }
+    // ...but notify the main Postgres DB's notification bus.
+    payload := constructPayload(event, key)
+    return b.conn.Bus().Notify(buildEventsNotificationChannel(b.ID()), payload)
+}
+```
+
+```go
+func (b *build) Events(ctx context.Context, from uint) (EventSource, error) {
+    notifier, err := newConditionNotifier(b.conn.Bus(), buildEventsNotificationChannel(b.ID()), func() (bool, error) {
+        return true, nil
+    })
+    if err != nil {
+        return nil, err
+    }
+
+    return newBuildEventSource(
+        ctx,
+        b,
+        b.conn,
+        b.eventStore, // fetch existing events from the EventStore using EventStore.Get(...)
+        notifier,     // but subscribe to the Postgres notification bus to receive new events
+        from,
+    ), nil
+}
+```
+
+
+## <a name="migrations">Migrations</a>
+
+Currently, we have migrations to set up the `build_events` table, as well as
+`TRIGGER`s for creating/deleting `pipeline_build_events_x` and
+`team_build_events_x` tables when pipelines and teams are created/deleted.
+
+I propose that we comment out these up migrations (so future deployments don't
+get these tables/triggers), and leave any database initialization (or whatever
+appropriate backend setup) to `BuildEvent.Setup()`. This means `Setup` will
+have to be "smart" (i.e. have to handle the cases where the initialization has
+already been applied). It also makes it difficult to make changes to the
+schema, unlike with migrations - each `EventStore` (that stores data in a
+structured way - this may just end up being Postgres) will need to basically
+implement its own form of migrations if we need to change the schema. My guess
+is that we won't be changing the schema too often, though (given that we
+haven't changed `build_events` since the initial migration).
+
+
+## Configuration
+
+One thing I haven't really explored is how operators would configure the
+backend store. I think it could be similar to how Credential managers are
+configured - there are groups of independent configuration options for each
+possible `EventStore`. If certain flags are configured for a given backend, use
+that backend.
```

Comment:
I think that it would be desirable to design build event stores in such a way that allows them to be built and released independently from the rest of Concourse, similar to how Terraform eventually spun providers out of its main source tree so that they could be built and released independently, typically written and maintained by the commercial entities backing the providers.

For example, a user might wish to export build logs to Coralogix, which is a log management and analysis platform. As Coralogix is a commercial platform, I would fully expect that a Coralogix build event store plugin would be written and maintained by Coralogix and not by the Concourse team / Pivotal.

(full disclosure: I work for Coralogix and we would probably (but I'm intentionally being non-committal here) be interested in building out a plugin for our own platform as part of dogfooding our own product)

Reply:
That's very interesting - I hadn't put much thought into an architecture like that, mostly because the precedent seems to be package and release everything from within Concourse (at least in terms of credential managers and metrics emitters). I think it'd be cool to for build event stores to be written as plugins external to the Concourse binary, and it'd be really cool to have experts in the domain be the ones to implement/maintain it.

I guess one concern is the added operational complexity, though I imagine it's not too bad. Like Terraform, we could package some plugins directly into Concourse (it could just be the Postgres plugin so operators don't need to bring their own plugin for the default behaviour).

Reply:
I was thinking about this a bit more. For *some* use cases of the proposed `EventStore`, having their implementations be invoked directly (as opposed to over something like gRPC) is preferable. For instance, invoking the Postgres `EventStore` in memory allows us to perform queries database objects (e.g. [also deleting all of a team's pipelines when deleting a team](https://github.com/concourse/concourse/pull/5651/commits/18832d57e1a175b6931ef7a845c01a87c11836e8#diff-35004ec688e559fe368187a2c0638fefR266)). For certain applications of `EventStore` composition (e.g. [for secret redaction](https://github.com/aoldershaw/rfcs/blob/config-event-store/053-configurable-event-store/proposal.md#eventstore-composition)), handling the operations in memory is beneficial.

So, what if `EventStore`s are just an in-memory type, but there's a `PluginEventStore` implementation that functions as a gRPC client? Which plugin (gRPC server) to use could be configured via flags to Concourse, and any environment variables such as `CONCOURSE_EVENTSTORE_PLUGIN_*` could be passed along to the plugin on startup to configure e.g. authentication (similar to how we support [custom configuration for Guardian](https://github.com/concourse/concourse/blob/72c7cdad6f2ef05036c6c9fb44a2fc5eee6cbb88/worker/workercmd/guardian.go#L116)).

What do you think of that approach?

Reply:
I doubt in-memory EventStore, because that implies every ATC will duplicated store build logs, which might not fit to huge clusters where hundreds pipelines generate huge amount of build logs. But I think plugin is a good idea, at least, a PG plugin would allow us to store build logs to a separate database, if the plugin is well designed, it may even support to dispatch build logs to multiple databases, which would definitely benefit scalability of Concourse. Also, users will have the flexibility to store build logs in various backends, ES, and whatever.